python-checks/src/main/java/org/sonar/python/checks/UndefinedNameAllPropertyCheck.java

```diff
@@ -0,0 +1,163 @@
+import org.sonar.python.semantic.BuiltinSymbols;
+import org.sonar.python.tree.TreeUtils;
+
+@Rule(key = "S5807")
+public class UndefinedNameAllPropertyCheck extends PythonSubscriptionCheck {
+
+  @Override
+  public void initialize(Context context) {
+    context.registerSyntaxNodeConsumer(Tree.Kind.ASSIGNMENT_STMT, ctx -> {
+      AssignmentStatement assignmentStatement = (AssignmentStatement) ctx.syntaxNode();
+      if (TreeUtils.firstAncestorOfKind(assignmentStatement, Tree.Kind.CLASSDEF, Tree.Kind.FUNCDEF) != null) {
+        // We only consider __all__ assignment at module level
+        return;
+      }
+      ExpressionList expressionList = assignmentStatement.lhsExpressions().get(0);
+      if (expressionList.expressions().size() > 1) {
+        return;
+      }
+      Expression lhs = expressionList.expressions().get(0);
+      if (lhs.is(Tree.Kind.NAME) && ((Name) lhs).name().equals("__all__")) {
+        checkAllProperty(ctx, assignmentStatement);
+      }
+    });
+  }
+
+  private static void checkAllProperty(SubscriptionContext ctx, AssignmentStatement assignmentStatement) {
+    Expression assignedValue = assignmentStatement.assignedValue();
+    if (!assignedValue.is(Tree.Kind.LIST_LITERAL) && !assignedValue.is(Tree.Kind.TUPLE)) {
+      return;
+    }
+    List<Tree> stringExpressions = getStringExpressions(assignedValue);
+    FileInput fileInput = (FileInput) TreeUtils.firstAncestorOfKind(assignedValue, Tree.Kind.FILE_INPUT);
+    if (fileInput == null) {
+      // Should never happen
+      return;
+    }
+    UnresolvedWildcardImportVisitor unresolvedWildcardImportVisitor = new UnresolvedWildcardImportVisitor();
+    fileInput.accept(unresolvedWildcardImportVisitor);
+    if (unresolvedWildcardImportVisitor.hasUnresolvedWildcardImport) {
+      return;
+    }
+    Map<String, Symbol> symbolsByName = fileInput.globalVariables().stream().collect(Collectors.toMap(Symbol::name, Function.identity()));
+    for (Tree stringExpression : stringExpressions) {
+      StringLiteral stringLiteral = retrieveStringLiteral(stringExpression);
+      if (isUnknownSymbol(ctx.pythonFile(), symbolsByName, stringLiteral)) {
+        PreciseIssue issue = ctx.addIssue(stringExpression, String.format("Change or remove this string; \"%s\" is not defined.", stringLiteral.trimmedQuotesValue()));
+        if (stringExpression != stringLiteral) {
+          issue.secondary(stringLiteral, null);
+        }
+      }
+    }
+  }
+
+  private static boolean isUnknownSymbol(PythonFile pythonFile, Map<String, Symbol> symbolsByName, StringLiteral stringLiteral) {
+    if (stringLiteral.stringElements().stream().anyMatch(StringElement::isInterpolated)) {
+      return false;
+    }
+    return !symbolsByName.containsKey(stringLiteral.trimmedQuotesValue())
+      && !BuiltinSymbols.all().contains(stringLiteral.trimmedQuotesValue())
+      && !isInitFileExportingModule(pythonFile, stringLiteral.trimmedQuotesValue());
```

Comment:
_subjective, minor_ It looks a little bit weird to me that three out of the four conditions are inlined into a single expression, but the first one has its own if statement. I would either have them all in the return or split them into if statements.